src/main/java/org/opentripplanner/updater/vehicle_sharing/vehicle_presence/CarPresencePredictor.java

```diff
@@ -0,0 +1,99 @@
+    private double length;
+
+    /**
+     * Calculate prediction value for vehicle presence in given position. vehiclePresencePredictor keeps cells with
+     * positive probability of presence in future so if no cell is found it assumes that vehicle disappears and returns 0.
+     *
+     * @param vehicleDescription vehicle which we want to make prediction about
+     * @param time               timestamp in seconds since epoch of point in time when we want to make prediction.
+     *                           Used for choosing time braked from [0min, 15min, 30min, 45min] in the future.
+     * @return probability value in range [0,1] indicating presence of vehicle in given time.
+     */
+    public double predict(VehicleDescription vehicleDescription, long time) {
+        if (VehicleType.CAR.name().equalsIgnoreCase(vehicleDescription.getVehicleType().name())) {
+            STRtree tree = chooseIndex(time);
+            if (tree == null) {
+                return 1;
+            } else {
+                return findGridCellValue(vehicleDescription.getLongitude(), vehicleDescription.getLatitude(), tree);
+            }
+        } else {
+            return 1;
+        }
+    }
+
+    public void updateVehiclePresenceHeatmap(VehiclePresence vehiclePresenceHeatmaps) {
+        if (vehiclePresenceHeatmaps.getVehicle().equalsIgnoreCase(VehicleType.CAR.name())) {
+            index15 = new STRtree();
+            for (VehiclePresence.Prediction prediction : vehiclePresenceHeatmaps.getPredictions_15()) {
+                index15.insert(prediction.getEnvelope(), prediction);
+            }
+            index30 = new STRtree();
+            for (VehiclePresence.Prediction prediction : vehiclePresenceHeatmaps.getPredictions_30()) {
+                index30.insert(prediction.getEnvelope(), prediction);
+            }
+            index45 = new STRtree();
+            for (VehiclePresence.Prediction prediction : vehiclePresenceHeatmaps.getPredictions_45()) {
+                index45.insert(prediction.getEnvelope(), prediction);
+            }
+            index15.build();
+            index30.build();
+            index45.build();
+            this.time = vehiclePresenceHeatmaps.getTimestamp();
+            this.length = vehiclePresenceHeatmaps.getCellLength();
+            this.width = vehiclePresenceHeatmaps.getCellWidth();
+        }
+    }
+
+    private double findGridCellValue(double longitude, double latitude, STRtree tree) {
+
+        List<VehiclePresence.Prediction> predictions = tree.query(new Envelope(longitude - width, longitude + width, latitude - length, latitude + length));
+        for (VehiclePresence.Prediction prediction : predictions) {
+            if (prediction.getEnvelope().contains(longitude, latitude)) {
+                return prediction.getValue();
+            }
+        }
+        return 0;
+    }
+
+    @Nullable
+    private STRtree chooseIndex(long time) {
+        Instant pointInTime = Instant.ofEpochSecond(time);
```

Comment:
This is generally really nice code, but I am afraid it could be potentially not efficient enough? Not sure though, as we have <5k vehicles in whole city. Maybe you could precalculate `this.time + 7*60+30`, `this.time + 22*60+30` and `this.time + 37*60+30` and compare it with argument using `<`? I would call those `indexThreshold15`, ... and store them as class fields

Reply:
I simplified it, but as for using `<` operator am I against. I bet JIT compiler will spiffily this part of code to simple operation and `.isAfter` is more descriptive.

Reply:
I agree with Kamyki